src/rebase.rs

```diff
@@ -0,0 +1,487 @@
+    pub fn operation_count(&self) -> usize {
+        unsafe { raw::git_rebase_operation_entrycount(self.raw) }
+    }
+
+    /// Aborts a rebase that is currently in progress,
+    /// resetting the repository and working directory to their state before rebase began.
+    pub fn abort(&self) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_rebase_abort(self.raw));
+        }
+        Ok(())
+    }
+
+    /// Commits the current patch.
+    ///
+    /// You must have resolved any conflicts that were introduced during the
+    /// patch application from the iteration over `operation_iter`/`next`
+    /// invocation.
+    pub fn commit(&self,
+                  author: Option<&Signature>,
+                  committer: &Signature,
+                  message: Option<&str>)
+                  -> Result<Oid, Error> {
+        let mut raw = raw::git_oid { id: [0; raw::GIT_OID_RAWSZ] };
+        let message = try!(::opt_cstr(message));
+        unsafe {
+            try_call!(raw::git_rebase_commit(&mut raw,
+                                             self.raw,
+                                             author.map(|s| s.raw() as *const _),
+                                             committer.raw(),
+                                             0 as *const c_char,
+                                             message));
+            Ok(Binding::from_raw(&raw as *const _))
+        }
+    }
+
+    /// Finishes a rebase that is currently in progress once all patches have been applied.
+    pub fn finish(&self, signature: Option<&Signature>) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_rebase_finish(
+                self.raw,
+                signature
+                    .map(|s| s.raw() as *const raw::git_signature)
+                    .unwrap_or(0 as *const raw::git_signature)
+            ));
+        }
+        Ok(())
+    }
+    /// Gets the index produced by the last operation.
+    /// This is useful for resolving conflicts in an in-memory rebase before committing them.
+    pub fn inmemory_index(&self) -> Result<Index, Error> {
+        let mut ret = 0 as *mut raw::git_index;
+        unsafe {
+            try_call!(raw::git_rebase_inmemory_index(&mut ret, self.raw));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Performs the next rebase operation and returns the `RebaseOperation` about it.
+    ///
+    /// If the operation is one that applies a patch (which is any operation
+    /// except GIT_REBASE_OPERATION_EXEC) then the patch will be applied and
+    /// the index and working directory will be updated with the changes. If
+    /// there are conflicts, you will need to address those before committing
+    /// the changes.
+    ///
+    /// Iterating over the result of `operations()` correctly handles begin and
+    /// end conditions for you.
+    pub fn next<'rebase>(&'rebase self) -> Result<RebaseOperation<'rebase, 'repo>, Error> {
```

Comment:
To avoid weird iterator issues perhaps this could take `&'rebase mut self`?

Reply:
I'm not sure exactly which weird iterator issues you're talking about, the main issue from my perspective isn't intra-process race conditions (which are weird) but the fact that rebases stop all the time and it's not crazily uncommon for rebases to start in one tool and then have some of the work happen in another tool.

For example, starting a rebase in an IDE, running into a conflict, editing in the terminal, then running `git rebase --continue` in the terminal, before jumping back into the IDE. There is a similar problem with the `operation_count`, inasmuch as it's possible for a rebase to be completed and a new one to be started which will cause unexpected errors. But in that case the entire rebase is invalid so I think it's a more reasonable to throw an error there.

I guess the other position is that keeping track of where the in-memory rebase is supposed to be allows the developer to decide how to handle the situation.

Also I can't seem to get the borrow checker to agree with addng that `mut`.

Reply:
I guess another way of saying that is that `Rebase` can't be immutable: outside processes can affect its state.

Additionally, `RebaseOperations` should probably be either fully stateless -- getting rid of the `operation_count` field, and not requiring uniq access. Or it should be stateful and I think everything would be easier if `fn operations()` took self by-value. I've implemented that as well as adding the `mut` to the borrow (which I never got working) let me know what you think.